server/service/service_osquery.go

```diff
@@ -419,11 +420,59 @@ const hostDistributedQueryPrefix = "fleet_distributed_query_"
 
 // detailQueriesForHost returns the map of detail+additional queries that should be executed by
 // osqueryd to fill in the host details.
+
+var (
+	lastCheck time.Time
+	rate      int = 10
+)
+
 func (svc *Service) detailQueriesForHost(ctx context.Context, host fleet.Host) (map[string]string, error) {
 	if !svc.shouldUpdate(host.DetailUpdatedAt, svc.config.Osquery.DetailUpdateInterval) && !host.RefetchRequested {
 		return nil, nil
 	}
 
+	// Update the rate every 5 seconds
+	if time.Since(lastCheck) > 5*time.Second {
+		lastCheck = time.Now()
+
+		hostStats, err := svc.ds.GenerateHostStatusStatistics(ctx, fleet.TeamFilter{User: &fleet.User{GlobalRole: ptr.String(fleet.RoleAdmin)}}, time.Now())
+		if err != nil {
+			return nil, ctxerr.Wrap(ctx, err, "get host stats for detail queries")
+		}
+
+		// Attempt to calculate a rate that allows all the online hosts to perform their update over
+		// a single interval.
+		rate = int(hostStats.OnlineCount/uint(svc.config.Osquery.DetailUpdateInterval.Seconds())) + 1
+		if rate < 10 {
+			rate = 10
+		}
+
+		fmt.Println(
+			"interval:", svc.config.Osquery.DetailUpdateInterval,
+			"online:", hostStats.OnlineCount,
+			"rate: ", rate,
+		)
+	}
+
+	limiter, err := throttled.NewGCRARateLimiter(
+		svc.limitStore,
+		throttled.RateQuota{MaxRate: throttled.PerSec(rate), MaxBurst: rate},
+	)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "rate limit detail queries")
+	}
+	// Maybe there should be a separate limit for if the refetch was requested explicitly? So that
+	// any hosts with refetch requested can do so quicker.
+	limited, _, err := limiter.RateLimit("detailQueries", 1)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "check rate limit detail queries")
```

Comment:
Qs: 
If redis:
- is having performance issues do we want to delay the distributed read request processing? (maybe ok)
- is down, do we want to fail the whole distributed read request? (not ok IMO)

Reply:
Great points. This could use some careful thought.

> is down, do we want to fail the whole distributed read request? (not ok IMO)

Agreed, rate limiting check failure should not fail the entire request.

> is having performance issues do we want to delay the distributed read request processing? (maybe ok)

I'm less sure on this -- If we delay sending the query it could help the infrastructure return to stability, however it could also prevent hosts from updating even when they don't really need functional Redis to do that.

Reply:
On #3318 I ignore rate limiting errors, and added a INFO log in case "rate limiting processing" takes more than 2 seconds.